### ISSUE TITLE
plugin: drop runtime from plugins

### ISSUE THREAD

**Issue by rasky** [golang/go#17150]
It looks like plugins contain a copy of the runtime. I guess that copy is unused, and could probably be dropped, reducing binary size.

**Comment by crawshaw** [comment]
Yup, I had a TODO about this I lost somewhere in the CL patchsets.

It should be straightforward, it's already possible to build -linkshared binaries without a runtime. I expect most of the work is careful argument construction in cmd/go.

**Comment by rsc** [comment]
@crawshaw, still want to do this for Go 1.8?

**Comment by crawshaw** [comment]
Pushing to 1.9, so I can spend what cycles I have on bugs.

**Comment by crawshaw** [comment]
@hirochachacha at one point you had a CL exploring removing the runtime from plugins. Are you still working on it?

**Comment by hirochachacha** [comment]
@crawshaw No, I'm afraid not.

**Comment by montao** [comment]
I think I have this issue too. My code gitlab.com/montao/forbasile

**Comment by bstarynk** [comment]
For what it is worth, my [manydl.c](https://github.com/bstarynk/misc-basile/blob/master/manydl.c) is generating "randomly" the code of many plugins in C and compiling it. That program demonstrates that on Linux, one can have hundreds of thousands of plugins. I see no reason why that would not be possible in Go (and I guess that writing the equivalent `manydl.go` program is reasonably easy)
And generating plugins at runtime is sometimes interesting.

So I believe it would be great if Go would practically support many plugins (in practice generated ones). The current implementation does that poorly (it is designed for a few, hand-written, plugins in mind).
In its current implementation, each generated `*.so` would have its own copy of common standard functions (e.g. `fmt.Printf`), so if we generate thousands of plugins, the binary code of `fmt.Printf` gets duplicated a thousand times in the virtual address space. That does not scale.

**Comment by montao** [comment]
> 
> 
> @hirochachacha at one point you had a CL exploring removing the runtime from plugins. Are you still working on it?

I found this CL, I think: https://go-review.googlesource.com/c/go/+/61090 Not sure what is the problem with it but it was abandoned. Any clues?

**Comment by hirochachacha** [comment]
I don't remember exactly, but I couldn't recommend investing that approach since that is a half-baked solution after all. It only cover `runtime` package, it doesn't cover other standard libraries. For example,  if there are two plugin A and B that use `strings` package, both plugins still need to bundle `strings` package. Ideally, this issue should be solved by re-implementing plugin system on top of `buildmode=shared`. I guess that it requires hard work though.
